File: etsyCatalogs/etsyCatalogs/View/CatalogsDatasource.m
//
//  CatalogsDatasource.m
//  etsyCatalogs
//
//  Created by Ihonahan Buitrago on 9/1/14.
//  Copyright (c) 2014 Ihonahan Buitrago. All rights reserved.
//

#import "CatalogsDatasource.h"


@interface CatalogsDatasource()

@property(nonatomic, strong) NSString *currentKeywords;
@property(nonatomic, assign) int currentPage;

@end


@implementation CatalogsDatasource


- (void)rewindList
{
    self.currentPage--;
    [self getListPage];
}

- (void)fastForwardList
{
    self.currentPage++;
    [self getListPage];
}

- (void)getInitialListForKeywords:(NSString *)keywords
{
    self.currentPage = 0;
    self.currentKeywords = keywords;
    [self getListPage];
}

- (void)getListPage
{
    [CatalogServices getCatalogListingsByWeywords:self.currentKeywords inPage:self.currentPage withBlock:^(CatalogResult *catalogResults, NSError *error) {
        if (error) {
            // Report the error
            if (self.delegate) {
                if ([self.delegate respondsToSelector:@selector(catalogsDatasource:DidFailWithError:)]) {
                    [self.delegate catalogsDatasource:self DidFailWithError:error];
                }
            }
        } else {
            // Everything went just OK, we reload the table and report to the delegate
            self.catalogResults = catalogResults;
            [self.listingsTable reloadData];
            if (self.delegate) {
                if ([self.delegate respondsToSelector:@selector(catalogsDatasource:DidLoadListingCatalogs:)]) {
                    [self.delegate catalogsDatasource:self DidLoadListingCatalogs:self.catalogResults.count];
                }
            }
        }
    }];
}



#pragma mark - Table View

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.catalogResults.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    ListingTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"ListingTableViewCell" forIndexPath:indexPath];
    
    if (cell == nil) {
        cell = [[ListingTableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"ListingTableViewCell"];
    }
    
    CatalogItem *catalogItem = [self.catalogResults getCatalogItemByIndex:indexPath.row];
    cell.summaryLabel.text = catalogItem.itemTitle;
    if (catalogItem.images.count) {
        CatalogListingImage *catalogImage = [catalogItem.images objectAtIndex:0];
        NSString *urlImg = catalogImage.url170x135;
        [cell.listingImage setImageWithURL:[NSURL URLWithString:urlImg]];
    } else {
        [CatalogServices getListingImagesForListingID:catalogItem.listingID withBlock:^(NSArray *imagesResult, NSError *error) {
            if (error) {
                // TODO: Do something with the error.
            } else {
                if (imagesResult.count) {
                    CatalogListingImage *img0 = [imagesResult objectAtIndex:0];
                    CatalogItem *itemFound = [self.catalogResults getCatalogItemWithListingID:img0.listingID];
                    if (itemFound) {
                        itemFound.images = [NSMutableArray arrayWithArray:imagesResult];
                    }
                }
            }
        }];
    }
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    ListingTableViewCell *cell = (ListingTableViewCell *)[tableView cellForRowAtIndexPath:indexPath];
    [cell setSelected:NO animated:NO];
}

@end

File: etsyCatalogsSwift/etsyCatalogsSwift/Model/CatalogsDataSource.swift
//
//  CatalogsDataSource.swift
//  etsyCatalogsSwift
//
//  Created by Ihonahan Buitrago on 9/19/14.
//  Copyright (c) 2014 Ihonahan Buitrago. All rights reserved.
//

import Foundation

import UIKit


protocol CatalogsDataSourceDelegate
{
    func catalogsDataSourceDidLoadListingCatalogs(catalogDataSource : CatalogsDataSource, count : Int)
    func catalogsDataSourceDidFailWithError(catalogDataSource : CatalogsDataSource, error : NSError?)
    func catalogsDataSourceSelectedDetailsForListingImage(catalogDataSource : CatalogsDataSource, catalogItem : CatalogItem)
}


class CatalogsDataSource: NSObject, UITableViewDataSource, UITableViewDelegate
{
    @IBOutlet var listingTable : UITableView!
    
    var delegate : CatalogsDataSourceDelegate!
    var catalogResults : CatalogResult = CatalogResult()
    
    var currentPage : Int = 0
    var currentKeywords : String = ""
    var placeholder : UIImage =  UIImage(named: "transparent-square-th.png")
    
    override init() {
        super.init()
    }


    func rewindPage() {
        if (isValidGoBackward()) {
            currentPage--
            getListPage()
        }
    }


    func isValidGoBackward() -> Bool {
        if (currentPage > 0) {
            return true
        } else {
            reportErrorToDelegate(nil)
            return false
        }
    }
    
    
    func fastForwardPage() {
        if (isValidGoForward()) {
            currentPage++
            getListPage()
        }
    }

    
    func isValidGoForward() -> Bool {
        if (catalogResults.pagination.nextPage > self.currentPage) {
            return true;
        }
        
        reportErrorToDelegate(nil)
        return false;
    }
    
    
    func getInitialListForKeywords(keywords : String) {
        self.currentPage = 0
        self.currentKeywords = keywords
        getListPage()
    }
    
    
    func getListPage() {
        self.listingTable.reloadData()

        CatalogServices.getCatalogListingsByKeywordsAndPage(self.currentKeywords,
            pageNumber: self.currentPage,
            block: { (catalogsResults, error) -> Void in
                if let givenError = error {
                    self.reportErrorToDelegate(error)
                } else {
                    if let givenResults = catalogsResults {
                        self.catalogResults = catalogsResults!
                        self.listingTable.reloadData()
                        
                        if let givenDelegate = self.delegate {
                            self.delegate.catalogsDataSourceDidLoadListingCatalogs(self, count: self.catalogResults.count())
                        }
                    }
                    
                }
        })
    }
    
    
    func reportErrorToDelegate(error: NSError?) {
        if let givenDelegate = self.delegate {
            self.delegate.catalogsDataSourceDidFailWithError(self, error: error)
        }
    }

    func handleDetail(sender : UIButton) {
        let row : Int = sender.tag
        
        let item : CatalogItem = self.catalogResults.getCatalogItemByIndex(row)!
        
        self.delegate?.catalogsDataSourceSelectedDetailsForListingImage(self, catalogItem: item)
    }
    
    func initialSetOfImagesForListingID(listingId : Int) {
        CatalogServices.getListingImagesForListingID(listingId,
            block: { (imagesResult, error) -> Void in
                if let givenError = error {
                    self.reportErrorToDelegate(error)
                } else {
                    if let givenArray : Array<CatalogListingImage> = imagesResult {
                        if givenArray.count > 0 {
                            let imageItem : CatalogListingImage = givenArray[0]
                            let catalogItem : CatalogItem = self.catalogResults.getCatalogItemWithListingID(imageItem.listingID!)!
                            catalogItem.images = givenArray
                            let row : Int = self.catalogResults.findIndexOfItem(catalogItem)
                            if row >= 0 {
                                let indexPath : NSIndexPath = NSIndexPath(forRow: row, inSection: 0)
                                
                                if let cell : UITableViewCell = self.listingTable.cellForRowAtIndexPath(indexPath) {
                                    var imagecell : UIImageView = cell.viewWithTag(110) as UIImageView
                                    imagecell.setImageWithURL(NSURL.URLWithString(imageItem.url170x135!),
                                        placeholderImage: self.placeholder)
                                }
                            }
                        }
                    }
                }
        })
    }
    
    
    // MARK: - UITableViewDatasource and Delegate methods
    func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if let results = self.catalogResults as CatalogResult? {
            return self.catalogResults.count()
        } else {
            return 0
        }
    }
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        var cell : UITableViewCell = tableView.dequeueReusableCellWithIdentifier("ListingTableViewCell", forIndexPath: indexPath) as UITableViewCell
        
        let titleCell : UILabel = cell.viewWithTag(100) as UILabel
        let imageCell : UIImageView = cell.viewWithTag(110) as UIImageView
        let buttonCell : UIButton = cell.viewWithTag(120) as UIButton
        
        buttonCell.tag = indexPath.row
        buttonCell.addTarget(self, action: "handleDetail:", forControlEvents: UIControlEvents.TouchUpInside)
        imageCell.image = self.placeholder
        
        if let item : CatalogItem = self.catalogResults.getCatalogItemByIndex(indexPath.row) {
            titleCell.text = item.itemTitle
            if item.images.count > 0 {
                var catalogImage : CatalogListingImage = item.images[0]
                let urlImg : String = catalogImage.url170x135!
                imageCell.setImageWithURL(NSURL.URLWithString(urlImg))
            } else {
                self.initialSetOfImagesForListingID(item.listingID!)
            }
        }
        
        return cell
    }
    
    func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
    
    }
    
    
    
}